main.go

```diff
@@ -14,7 +14,291 @@
+			),
+			xconnect.NewServer(vppConn),
+			acl.NewServer(vppConn, config.ACLConfig),
+			mechanisms.NewServer(map[string]networkservice.NetworkServiceServer{
+				memif.MECHANISM: chain.NewNetworkServiceServer(memif.NewServer(vppConn)),
+			}),
+			connect.NewServer(
+				ctx,
+				client.NewClientFactory(
+					client.WithName(config.Name),
+					client.WithAdditionalFunctionality(
+						metadata.NewClient(),
+						mechanismtranslation.NewClient(),
+						replacelabels.NewClient(config.Labels),
+						up.NewClient(ctx, vppConn),
+						xconnect.NewClient(vppConn),
+						memif.NewClient(vppConn),
+						sendfd.NewClient(),
+						recvfd.NewClient(),
+					)),
+				connect.WithDialOptions(clientOptions...),
+			),
+		))
+	// ********************************************************************************
+	log.FromContext(ctx).Infof("executing phase 5: create grpc server and register firewall-server")
+	// ********************************************************************************
+	server := grpc.NewServer(append(
+		opentracing.WithTracing(),
+		grpc.Creds(
+			grpcfd.TransportCredentials(
+				credentials.NewTLS(
+					tlsconfig.MTLSServerConfig(source, source, tlsconfig.AuthorizeAny()),
+				),
+			),
+		),
+	)...)
+	firewallEndpoint.Register(server)
+
+	tmpDir, err := ioutil.TempDir("", config.Name)
+	if err != nil {
+		logrus.Fatalf("error creating tmpDir %+v", err)
+	}
+	defer func(tmpDir string) { _ = os.Remove(tmpDir) }(tmpDir)
+	listenOn := &(url.URL{Scheme: "unix", Path: filepath.Join(tmpDir, config.ListenOn)})
+
+	srvErrCh := grpcutils.ListenAndServe(ctx, listenOn, server)
+	exitOnErr(ctx, cancel, srvErrCh)
+	log.FromContext(ctx).Infof("grpc server started")
+
+	// ********************************************************************************
+	log.FromContext(ctx).Infof("executing phase 6: register nse with nsm")
+	// ********************************************************************************
+
+	nseRegistryClient := registryclient.NewNetworkServiceEndpointRegistryClient(ctx, &config.ConnectTo, registryclient.WithDialOptions(clientOptions...))
+	nse, err := nseRegistryClient.Register(ctx, &registryapi.NetworkServiceEndpoint{
+		Name:                config.Name,
+		NetworkServiceNames: []string{config.ServiceName},
+		NetworkServiceLabels: map[string]*registryapi.NetworkServiceLabels{
+			config.ServiceName: {
+				Labels: config.Labels,
+			},
+		},
+		Url: listenOn.String(),
+	})
+	logrus.Infof("nse: %+v", nse)
+
+	if err != nil {
+		log.FromContext(ctx).Fatalf("unable to register nse %+v", err)
+	}
+
+	// ********************************************************************************
+	log.FromContext(ctx).Infof("startup completed in %v", time.Since(starttime))
+	// ********************************************************************************
+
+	// wait for server to exit
+	<-ctx.Done()
+	<-vppErrCh
+}
+
+func exitOnErr(ctx context.Context, cancel context.CancelFunc, errCh <-chan error) {
```

Comment:
@edwarnicke I see this function in many places. Can we move it into sdk/tools?

This can be done in separate PRs.